autopush/router/apns2.py

```diff
@@ -0,0 +1,148 @@
+import json
+from collections import deque
+
+import hyper.tls
+from hyper import HTTP20Connection
+from hyper.http20.exceptions import HTTP20Error
+
+from autopush.router.interface import RouterException
+
+SANDBOX = 'api.development.push.apple.com'
+SERVER = 'api.push.apple.com'
+
+APNS_PRIORITY_IMMEDIATE = 10
+APNS_PRIORITY_LOW = 5
+APNS_MAX_CONNECTIONS = 20
+
+
+class APNSException(Exception):
+    pass
+
+
+class APNSClient(object):
+    def __init__(self, cert_file, key_file, topic,
+                 alt=False, use_sandbox=False,
+                 max_connections=APNS_MAX_CONNECTIONS,
+                 logger=None, metrics=None):
+        """Create the APNS client connector.
+
+        The cert_file and key_file can be derived from the exported `.p12`
+        **Apple Push Services: *bundleID*** key contained in the **Keychain
+        Access** application. To extract the proper PEM formatted data, you
+        can use the following commands:
+
+        ```
+        openssl pkcs12 -in file.p12 -out apns_cert.pem -clcerts -nokeys
+        openssl pkcs12 -in file.p12 -out apns_key.pem -nocerts -nodes
+        ```
+
+        The *topic* is the Bundle ID of the bridge recipient iOS application.
+        Since the cert needs to be tied directly to an application, the topic
+        is usually similar to "com.example.MyApplication".
+
+        :param cert_file: Path to the PEM formatted APNs certification file.
+        :type cert_file: str
+        :param key_file: Path to the PEM formatted APNs key file.
+        :type key_file: str
+        :param topic: The *Bundle ID* that identifies the assoc. iOS app.
+        :type topic: str
+        :param alt: Use the alternate APNs publication port (if 443 is blocked)
+        :type alt: bool
+        :param use_sandbox: Use the development sandbox
+        :type use_sandbox: bool
+        :param max_connections: Max number of pooled connections to use
+        :type max_connections: int
+        :param logger: Status logger
+        :type logger: logger
+        :param metrics: Metric recorder
+        :type metrics: autopush.metrics.IMetric
+        """
+        self.server = SANDBOX if use_sandbox else SERVER
+        self.port = 2197 if alt else 443
+        self.ssl_context = hyper.tls.init_context(cert=(cert_file, key_file))
+        self.log = logger
+        self.metrics = metrics
+        self.topic = topic
+        self._max_connections = max_connections
+        self.connections = deque(maxlen=max_connections)
+        if self.log:
+            self.log.info("Starting APNS connection")
+
+        for i in xrange(0, max_connections):
```

Comment:
can we appendleft here just to be consistent w/ return?

in fact we can extendleft with a generator statement instead of a for loop (I loved extending).. and let's just use range() instead of xrange #barelypython3compat